prov/psm2/src/psmx2_init.c

```diff
@@ -458,8 +458,20 @@ static void psmx2_fini(void)
 {
 	FI_INFO(&psmx2_prov, FI_LOG_CORE, "\n");
 
-	if (! --psmx2_init_count)
-		psm2_finalize();
+	if (! --psmx2_init_count) {
+		/* This function is called from a library destructor, which is called
+		 * automatically when exit() is called. The call to psm2_finalize()
+		 * might cause deadlock if the applicaiton is terminated with Ctrl-C
+		 * -- the application could be inside a PSM call, holding a lock that
+		 * psm2_finalize() tries to acquire. This can be avoided by only
+		 * calling psm2_finalize() when PSM is guaranteed to be unused.
```

Comment:
Why doesn't this just result in psm2_finalize() waiting to acquire the lock? How does the psm2 library itself handle this scenario?